### PR TITLE
Add validation endpoints

### PR DESCRIPTION
**Pull request by VolhaAdaska**

### CODE REVIEW

File: GoogleSheetI18n.Api.SimpleWebApi/ClientApp/package.json [hunk 1]
```diff
@@ -39,7 +43,8 @@
     "build": "react-scripts build",
     "test": "cross-env CI=true react-scripts test --env=jsdom",
     "eject": "react-scripts eject",
-    "lint": "eslint ./src/"
+    "lint": "eslint ./src/",
+    "babel": "./node_modules/.bin/babel --extensions .babelrc src/**/*.{js,jsx,ts,tsx} --source-maps inline"
```
Comment (akhlebus) [inline GoogleSheetI18n.Api.SimpleWebApi/ClientApp/package.json:47]: Could you rename it to something more meaningful.

File: GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/components/Administration.js [hunk 2]
```diff
@@ -5,51 +5,75 @@ import { useTranslation } from 'react-i18next';
...
+    return (
+        <div>
+            <h1>{t('admin-page:administration')}</h1>
+            <h3>{t('admin-page:basic-actions')}</h3>
+            {isLoaded && <div>
+                <a href={linkToGoogleSheet} target="_blank" rel="noreferrer" className="btn btn-lg btn-success mb-3 mr-2">{t('admin-page:view-google-sheet')}</a>
+                {authState.isAuthenticated && <button onClick={clearCache} className="btn btn-lg btn-outline-secondary mb-3">
+                    {t('admin-page:reload-local-store')} </button>}
+                {authState.isAuthenticated && <button onClick={() => validateSpreadsheet(spreadsheetTypes.LOCAL)} className="btn btn-lg btn-outline-secondary mb-3 btn-validation">
+                    {t('admin-page:validation-local-store')} </button>}
```
Comment (akhlebus) [inline GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/components/Administration.js]: Could you change button name from `validation` to `validate`, in other words using verbs not nouns for button names.

File: GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/custom.css [hunk 3]
```diff
@@ -12,3 +12,7 @@ code {
   background-color: #1b6ec2;
   border-color: #1861ac;
 }
+
+.btn-validation {
```
Comment (akhlebus) [inline GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/custom.css]: Pls, remove it as we can utilize default bootstrap classes.

File: GoogleSheetI18n/Core/I18nGoogleClient.cs [hunk 4]
```diff
@@ -29,6 +33,33 @@ public async Task<IList<I18nSheet>> GetSheets(string spreadsheetId)
                 .ToArray();
         }
 
+        public string GetExtractedTranslations(string pageName, List<object> possibleKeys)
```
Comment (akhlebus) [inline GoogleSheetI18n/Core/I18nGoogleClient.cs]: Could you move the method to separate class `CodeTranslations`. And move validation logic to separate class.
In other words, it is better to separate retrieving data and validation logics.

File: GoogleSheetI18n/Core/I18nLocalStore.cs [hunk 5]
```diff
@@ -12,7 +12,7 @@ public class I18nLocalStore : II18nLocalStore
 
         public I18nLocalStore(string localStoreFolderPath)
         {
-            if (localStoreFolderPath is null or "")
+            if (string.IsNullOrEmpty(localStoreFolderPath))
```
Comment (akhlebus) [inline GoogleSheetI18n/Core/I18nLocalStore.cs]: I prefer the previous way as it is easier to read.

File: GoogleSheetI18n/Core/I18nLocalStore.cs [hunk 6]
```diff
@@ -21,18 +21,18 @@ public I18nLocalStore(string localStoreFolderPath)
                 ? $@"{AppDomain.CurrentDomain.BaseDirectory}\{localStoreFolderPath}"
                 : localStoreFolderPath;
 
-            if (localStoreFolderPath is not null and not "")
+            if (!string.IsNullOrEmpty(localStoreFolderPath))
```
Comment (akhlebus) [inline GoogleSheetI18n/Core/I18nLocalStore.cs:24]: The same as above.

File: GoogleSheetI18n/Core/I18nLocalStore.cs [hunk 7]
```diff
@@ -21,18 +21,18 @@ public I18nLocalStore(string localStoreFolderPath)
...
             {
                 Directory.CreateDirectory(_localStoreFolderPath);
             }
         }
 
         public async Task<I18nSheet> GetSheet(string spreadsheetId, string sheetName)
         {
-            if (spreadsheetId is null or "")
+            if (string.IsNullOrEmpty(spreadsheetId))
```
Comment (akhlebus) [inline GoogleSheetI18n/Core/I18nLocalStore.cs]: The same as above.

File: GoogleSheetI18n/Core/I18nLocalStore.cs [hunk 8]
```diff
@@ -21,18 +21,18 @@ public I18nLocalStore(string localStoreFolderPath)
...
                 throw new ArgumentNullException(nameof(spreadsheetId));
 
-            if (sheetName is null or "")
+            if (string.IsNullOrEmpty(sheetName))
```
Comment (akhlebus) [inline GoogleSheetI18n/Core/I18nLocalStore.cs]: The same as above.

File: GoogleSheetI18n/Core/I18nLocalStore.cs [hunk 9]
```diff
@@ -64,5 +64,20 @@ private string GetLocalStoreKey(string spreadsheetId, string sheetName)
         {
             return $"{spreadsheetId}.{sheetName}.json";
         }
+
+        public async Task<IList<I18nSheet>> GetSheets(string spreadsheetId)
+        {
+            var sheets = new List<I18nSheet>();
+            DirectoryInfo dir = new DirectoryInfo(_localStoreFolderPath);
+            FileInfo[] filesInDir = dir.GetFiles("*" + spreadsheetId + "*.*");
```
Comment (akhlebus) [inline GoogleSheetI18n/Core/I18nLocalStore.cs]: You can safely specify `json` extension here.

File: GoogleSheetI18n/Integrations/Web/I18nController.cs [hunk 10]
```diff
@@ -86,5 +101,51 @@ public object Get()
                 _i18nOptions.SubscriptionUrl
             };
         }
+
+        private object GetWarnings(IList<I18nSheet> i18nSheets)
```
Comment (akhlebus) [inline GoogleSheetI18n/Integrations/Web/I18nController.cs]: Looks like, this method shouldn't be here.

File: GoogleSheetI18n/Integrations/Web/I18nController.cs [hunk 11]
```diff
@@ -86,5 +101,51 @@ public object Get()
...
+                }
+            }
+
+            return new
+            {
+                Message = CreateErrorMessage(warnings)
+            };
+        }
+
+        private string CreateErrorMessage(List<string> warnings)
```
Comment (akhlebus) [inline GoogleSheetI18n/Integrations/Web/I18nController.cs]: Looks like, this method shouldn't be here.

**Comment by akhlebus** [comment]
Please, add some tests for new code.

File: GoogleSheetI18n.Api.SimpleWebApi/ClientApp/package.json (hunk 1 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n.Api.SimpleWebApi/ClientApp/package.json:47]: suggest something

File: GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/components/Administration.js (hunk 2 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/components/Administration.js]: renamed

File: GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/custom.css (hunk 3 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n.Api.SimpleWebApi/ClientApp/src/custom.css]: removed

File: GoogleSheetI18n/Core/I18nGoogleClient.cs (hunk 4 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n/Core/I18nGoogleClient.cs]: added new class i18nCodeTranslation. also the logic was split

File: GoogleSheetI18n/Integrations/Web/I18nController.cs (hunk 10 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n/Integrations/Web/I18nController.cs]: removed

File: GoogleSheetI18n/Integrations/Web/I18nController.cs (hunk 11 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n/Integrations/Web/I18nController.cs]: removed

File: GoogleSheetI18n/Core/I18nLocalStore.cs (hunk 9 above)
Comment (VolhaAdaska) [reply GoogleSheetI18n/Core/I18nLocalStore.cs]: used  json

**Review by akhlebus — APPROVED** [review]

**Pull request merged**